Graphql api docs (#12)
boostercloud/docs-site: source/includes/_application-api.md

# Your application API

> The API is the way the outside world interact with your system, and any server-side application needs one.
> With Booster, you don't need to worry about it: _it is created for you_ and evolves with your code _automatically_.

Reviewer: It's just my opinion, but it doesn't look like this sentence is saying anything useful to me. It's trying to sell me something, although I'm already reading the docs. I'd rephrase it as:

```suggestion
The API is generated automatically when you deploy your application, and it changes when needed, no additional steps required on your side.
```

When you finish deploying your application using the `boost deploy` command, it will print useful information needed to 
use your API. It will be shown under the "Outputs" section.

```sh
Outputs:
<application name>.httpURL = https://<API ID>.execute-api.<region>.amazonaws.com/<environment name>
<application name>.websocketURL = wss://<API ID>.execute-api.<region>.amazonaws.com/<environment name>
<application name>.clientID = abcdXjk1234

> ```

Reviewer: Note that this is explaining how it works in AWS. Another provider might not have this code. I'd remove this and say that after the application is deployed, the endpoints will be shown

Author: I agree, but let's do it in another PR in the new docs, just to merge this and move the docs there

The meaning of those values are:

> - **httpURL**: This is the main URL of your application. You will need it to interacting with the 

Reviewer 2: ```suggestion
- **httpURL**: This is the base HTTP URL of your application. You will need it to interacting with the 
```

```http request
POST https://<httpURL>/auth/sign-up
```
#### Request body
```json
{
	"clientId": "string",
	"username": "string",
	"password": "string",
	"userAttributes": {
   		"roles": ["string"]
	}
}
```

Parameter | Description
--------- | -----------
_clientId_ | The application client Id that you got as an output when the application was deployed.
_username_ | The username of the user you want to register. It **must be an email**.
_password_ | The password the user will use to later login into your application and get access tokens.
_userAttributes_ | Here you can specify the attributes of your user. These are: <br/> - _roles_  An array of roles this user will have. You can only specify here roles with the property `allowSelfSignUp = true`

#### Response
An empty body

#### Errors
> Sign-up error response body example: Not specifying an email as username.

```json
{
    "__type": "InvalidParameterException",
    "message": "Username should be an email."
}
```

You will get an HTTP status code different from 2XX and a body with a message telling you the reason of the error.

### Sign-in
Allows your users to get tokens to be able to make request to restricted endpoints.

> Remember that before a user can be signed in into your application, **its email must be confirmed**

Reviewer: ```suggestion
Remember that before a user can be signed in into your application, **their email must be confirmed**
```

#### Endpoint
```http request
POST https://<httpURL>/auth/sign-in
```
#### Request body
```json
{
	"clientId":"string",
	"username":"string",
	"password":"string"
}
```

Parameter | Description
--------- | -----------
_clientId_ | The application client Id that you got as an output when the application was deployed.

> _username_ | The username of the user you want to sign in. It must be previously signed up

Reviewer: ```suggestion
_username_ | The username of the user you want to sign in. They must have previously signed up
```

```json
{
    "__type": "UserNotConfirmedException",
    "message": "User is not confirmed."
}
```

You will get a HTTP status code different from 2XX and a body with a message telling you the reason of the error.

### Sign-out
Finalizes the user session by cancelling their tokens.

###### Endpoint
```http request
POST https://<httpURL>/auth/sign-out
```
###### Request body
> Sign-out request body

```json
{
	"accessToken":"string"
}
```

Parameter | Description
--------- | -----------
_accessToken_ | The access token you get in the sign-in call.

###### Response
An empty body
###### Errors
> Sign-out error response body example: Invalid access token specified

```json
{
    "__type": "NotAuthorizedException",
    "message": "Invalid Access Token"
}
```

You will get a HTTP status code different from 2XX and a body with a message telling you the reason of the error.

 

## Commands and ReadModels API
 
This is the main API of your application, as it allows you to:

 - _Modify_ data by **sending commands**
 - _Read_ data by **querying read models**
 - _Receive data in real time_ by **subscribing to read models** 
 

> All this is done through [GraphQL](https://graphql.org/), a nice query language for APIs that has useful advantages over simple REST APIs.

Reviewer: Let's make docs un-opinionated

```suggestion
All this is done through [GraphQL](https://graphql.org/), a query language for APIs that has useful advantages over simple REST APIs.
```

```sh
 wscat -c <websocketURL>
```

> Sending a message with the subscription

```json
{"query": "subscription { CartReadModel(id:\"demo\") { id items } }" }
```

After a successful subscription, you won't receive anything in return. Now, every time the read model you subscribed to
is modified, a new incoming message will appear in the socket with the updated version of the read model. This message
will have exactly the same format as if you were done a query with the same parameters.

Following the previous example, we now send a command (using a "mutation" operation) that adds
a new item with sku "ABC_02" to the `CartReadModel`. After it has been added, we receive the updated version of the read model through the
socket.

> Send the Command

```
URL: "<httpURL>/graphql"
```
```graphql
mutation {
  ChangeCart(input: {
    cartId: "demo"
    sku: "ABC_02"
    quantity: 3
  })
}
```

> The following message appears in the socket

```json
{
  "data": {
    "CartReadModel": {
      "id": "demo",
      "items": [
        {
          "sku": "ABC_01",
          "quantity": 2
        },
        {
          "sku": "ABC_02",
          "quantity": 3
        }
      ]
    }
  }
}

> ```

Reviewer 2: I think that it could be a good idea to show how to do this with ApolloClient in addition to the raw requests to show more realistic usage beyond pure documentation. We can do that in the new doc pages, no need to add it in this PR. /cc @NickSeagull @itrion

Author: Totally. I left that for the future because we still need to ensure Booster is compatible with the subscriptions-transport-ws to get full compatibility with Apollo

Author: BTW, feel free to merge the PR if you need that!
